File: android/app/src/main/kotlin/com/brunoouza/ar_tour/models/RecognitionCandidate.kt
package com.brunoouza.ar_tour.models

/**
 * Candidato a reconhecimento no pipeline híbrido.
 *
 * Contém o identificador do ponto e os scores parciais de cada componente,
 * além do score final calculado pelo [RecognitionFusionManager].
 *
 * @param pointId        ID do ponto (ex: "point_001")
 * @param pointName      Nome legível para logs e UI
 * @param latitude       Coordenada do ponto
 * @param longitude      Coordenada do ponto
 * @param imageReference Referência do marker ARCore (ex: "marker_01")
 * @param geoScore       Score de proximidade [0.0, 1.0]
 * @param markerScore    Score do marker ARCore [0.0, 1.0]
 * @param visualScore    Score visual OpenCV/ORB [0.0, 1.0] — 0.0 enquanto não integrado
 * @param finalScore     Score ponderado final calculado pelo FusionManager
 */
data class RecognitionCandidate(
    val pointId: String,
    val pointName: String,
    val latitude: Double,
    val longitude: Double,
    val imageReference: String,
    val geoScore: Float = 0f,
    val markerScore: Float = 0f,
    val visualScore: Float = 0f,
    val finalScore: Float = 0f
) {
    /** Converte para Map para serialização no EventChannel. */
    fun toMap(): Map<String, Any> = mapOf(
        "pointId"        to pointId,
        "pointName"      to pointName,
        "geoScore"       to geoScore,
        "markerScore"    to markerScore,
        "visualScore"    to visualScore,
        "finalScore"     to finalScore
    )
}

File: lib/features/recognition/domain/recognition_candidate.dart
import '../../../data/models/point_model.dart';

/// Candidato a ser reconhecido no pipeline híbrido.
///
/// Contém o ponto e os scores parciais de cada componente,
/// além do score final calculado pelo [RecognitionFusionManager].
class RecognitionCandidate {
  final PointModel point;

  /// Score de proximidade geográfica [0.0, 1.0].
  /// 1.0 = usuário está exatamente sobre o ponto; 0.0 = fora do raio.
  final double geoScore;

  /// Score do marker ARCore [0.0, 1.0].
  /// 1.0 = marker detectado em TrackingState.TRACKING.
  final double markerScore;

  /// Score visual por ORB/OpenCV [0.0, 1.0].
  /// 0.0 enquanto OpenCV não estiver integrado.
  final double visualScore;

  /// Score final ponderado: marker*0.55 + geo*0.20 + visual*0.25
  final double finalScore;

  const RecognitionCandidate({
    required this.point,
    this.geoScore = 0.0,
    this.markerScore = 0.0,
    this.visualScore = 0.0,
    this.finalScore = 0.0,
  });

  RecognitionCandidate copyWith({
    double? geoScore,
    double? markerScore,
    double? visualScore,
    double? finalScore,
  }) =>
      RecognitionCandidate(
        point: point,
        geoScore: geoScore ?? this.geoScore,
        markerScore: markerScore ?? this.markerScore,
        visualScore: visualScore ?? this.visualScore,
        finalScore: finalScore ?? this.finalScore,
      );

  @override
  String toString() =>
      'RecognitionCandidate(${point.id}, geo=$geoScore, marker=$markerScore, '
      'visual=$visualScore, final=$finalScore)';
}
